rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -38,21 +38,42 @@ private module Input1 implements InputSig1<Location> {
     }
   }
 
-  class TypeParameterPosition = TypeParam;
+  private newtype TTypeParameterPosition =
+    TTypeParamTypeParameterPosition(TypeParam tp) or
+    TSelfTypeParameterPosition()
+
+  class TypeParameterPosition extends TTypeParameterPosition {
+    TypeParam asTypeParam() { this = TTypeParamTypeParameterPosition(result) }
+
+    predicate isSelf() { this = TSelfTypeParameterPosition() }
+
+    string toString() {
+      result = this.asTypeParam().toString()
+      or
+      result = "Self" and this.isSelf()
+    }
+  }
+
+  /** Holds if `typeParam`, `param` and `ppos` all concern the same `TypeParam`. */
+  additional predicate typeParamMatchPosition(
+    TypeParam typeParam, TypeParamTypeParameter param, TypeParameterPosition ppos
+  ) {
+    typeParam = param.getTypeParam() and typeParam = ppos.asTypeParam()
+  }
 
   bindingset[apos]
   bindingset[ppos]
   predicate typeArgumentParameterPositionMatch(TypeArgumentPosition apos, TypeParameterPosition ppos) {
-    apos.asTypeParam() = ppos
+    apos.asTypeParam() = ppos.asTypeParam()
     or
-    apos.asMethodTypeArgumentPosition() = ppos.getPosition()
+    apos.asMethodTypeArgumentPosition() = ppos.asTypeParam().getPosition()
   }
 
-  private predicate id(Raw::TypeParam x, Raw::TypeParam y) { x = y }
+  private predicate id(Raw::AstNode x, Raw::AstNode y) { x = y }
 
-  private predicate idOfRaw(Raw::TypeParam x, int y) = equivalenceRelation(id/2)(x, y)
+  private predicate idOfRaw(Raw::AstNode x, int y) = equivalenceRelation(id/2)(x, y)
 
-  private int idOf(TypeParam node) { idOfRaw(Synth::convertAstNodeToRaw(node), result) }
+  private int idOf(AstNode node) { idOfRaw(Synth::convertAstNodeToRaw(node), result) }
```

Comment:
I think, instead of using `AstNode`, it would be better to use something like
```ql
private import codeql.rust.elements.internal.generated.Synth

private class TTypeParamOrTrait = Synth::TTypeParam or Synth::TTrait;

private class TypeParamOrTrait extends AstNode, TTypeParamOrTrait { }
```

in order to reduce pressure on the `equivalenceRelation` HOP.

Reply:
That makes sense. I remember that we also do `equivalenceRelation` on `Raw::AstNode` over in [the CFG implementation](https://github.com/github/codeql/blob/e6cf737f99e1af54691ac9044e453361ffb8cfbf/rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll#L56). Perhaps it would make sense to factor that one out and reuse it for both CFG and type inference?

Reply:
In order to be properly shared, we would have to make it cached, I'm not sure we'd want that.

Reply:
Is it only the type of `idOf` that should be changed? Not `idOfRaw` and `id`?

Reply:
I've changed `idOf` now. I picked at different name than `TypeParamOrTrait` as we'll also have to add `TypeAlias`es later when they become type parameters.

Reply:
I'm not sure how they should be updated to match the suggestion? I've changed `idOfRaw` now such that it only applies to the relevant extensional predicates, but it doesn't look like the above.